internal/jujuapi/http_proxy.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2024 Canonical.
+package jujuapi
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	"github.com/juju/names/v5"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/rpc"
+)
+
+type httpProxier struct {
+	jimm JIMM // interface
+}
+
+func (s *httpProxier) Authenticate(ctx context.Context, w http.ResponseWriter, req *http.Request) error {
+	// extract auth token
+	_, password, ok := req.BasicAuth()
+	if !ok {
+		return errors.E(errors.CodeUnauthorized, "authentication missing")
+	}
+	user, err := s.jimm.LoginWithSessionToken(ctx, password)
+	if err != nil {
+		return errors.E(errors.CodeUnauthorized, err)
+	}
+	// extract model uuid anche check permission
+	sPath, _ := strings.CutPrefix(req.URL.EscapedPath(), "/model")
+	uuid, _, err := modelInfoFromPath(sPath)
+	if err != nil {
+		return errors.E(errors.CodeUnauthorized, "cannot parse path")
+	}
+	access, err := s.jimm.GetUserModelAccess(ctx, user, names.NewModelTag(uuid))
+	if err != nil || (access != "admin" && access != "writer") {
+		return errors.E(errors.CodeUnauthorized, "unauthorized")
+	}
+	return nil
+}
+
+func (s *httpProxier) ServeHTTP(ctx context.Context, w http.ResponseWriter, req *http.Request) {
+	writeError := func(msg string, code int) {
+		w.WriteHeader(code)
+		_, err := w.Write([]byte(msg))
+		if err != nil {
+			zapctx.Error(ctx, "cannot write to connection", zap.Error(err))
+		}
+	}
+	sPath, _ := strings.CutPrefix(req.URL.EscapedPath(), "/model")
+	uuid, _, err := modelInfoFromPath(sPath)
```

Comment:
We don't need to extract the path segment because the full path is in the request?

Reply:
see https://github.com/canonical/jimm/pull/1357#discussion_r1751582926